### ISSUE TITLE
[glyf] bbox of transformed & anchored component calculated incorrectly

### ISSUE THREAD

**Issue by blueshade7** [fonttools/fonttools#1556]
The attached ttx file of a TrueType font (Variable, but this seems irrelevant) has its “O” glyph composed of a pair of “C” glyphs, one rotated by 180 degrees lined up to the other using anchor points (i.e., flag ARGS_ARE_XY_VALUES unset). The XML description of “O” glyph reads:
```
<TTGlyph name="O" xMin="56" yMin="-12" xMax="612" yMax="672">
  <component glyphName="C" x="0" y="0" flags="0x204"/>
  <component glyphName="C" firstPt="0" secondPt="9" scale="-1.0" flags="0x4"/>
</TTGlyph>
```
When converted to ttf by ttx then back to ttx, the XML description becomes:
```
<TTGlyph name="O" xMin="-334" yMin="-1356" xMax="334" yMax="672">
  <component glyphName="C" x="0" y="0" flags="0x204"/>
  <component glyphName="C" firstPt="0" secondPt="9" scale="-1.0" flags="0x4"/>
</TTGlyph>
```
The glyph bounding box is calculated incorrectly.
It appears as if translation offset for the component is calculated before rotation. These should be done backwards.

[SourceSansVariable-Roman.anchor.ttx.zip](https://github.com/fonttools/fonttools/files/3009833/SourceSansVariable-Roman.anchor.ttx.zip)

**Comment by anthrotype** [comment]
this is the code that computes the coordinates for a glyph component:

https://github.com/fonttools/fonttools/blob/c6006a7f8c31cd8e7dddad58894e1ba6c87e1362/Lib/fontTools/ttLib/tables/_g_l_y_f.py#L843-L874

By reading the description of SCALED_COMPONENT_OFFSET and UNSCALED_COMPONENT_OFFSET flags in https://docs.microsoft.com/en-us/typography/opentype/spec/glyf, I am not sure whether these flags apply also to the case when a component  args are not XY values but point numbers.

fonttools defaults to UNSCALED_COMPONENT_OFFSET, i.e. the component is fist scaled/rotated and then moved.

**Comment by justvanrossum** [comment]
If I build the font from the provided ttx I get a font that fonttools itself can't process:

```python
Traceback (most recent call last):
  ...
  File "fontTools/ttLib/ttFont.py", line 774, in draw
  File "fontTools/ttLib/tables/_g_l_y_f.py", line 1002, in draw
  File "fontTools/ttLib/tables/_g_l_y_f.py", line 1096, in getComponentInfo
AttributeError: 'GlyphComponent' object has no attribute 'x'
```

**Comment by anthrotype** [comment]
yep

https://github.com/fonttools/fonttools/blob/c6006a7f8c31cd8e7dddad58894e1ba6c87e1362/Lib/fontTools/ttLib/tables/_g_l_y_f.py#L1094-L1100

**Comment by justvanrossum** [comment]
And that comment is ... 20 years old :)

![image](https://user-images.githubusercontent.com/4246121/55077006-0ecb5080-5097-11e9-9a2a-6f9ae8f6bdea.png)

The number of fonts out there that use this feature must be rather small.

**Comment by anthrotype** [comment]
ok so, there are two separate issues here:
1) getComponentInfo doesn't support components with point numbers (so a glyph containing those can't be drawn using draw/drawPoints api)
2) the bbox is computed incorrectly because the Glyph.getCoordinates method is computing the offset (firstPoint - secondPoint) before applying the transformation, and then moving the transformed component by the unscaled offset.

The fix would be to compute the offset from the points only after the transform has been applied. Correct?

**Comment by justvanrossum** [comment]
> The fix for issue 2) would be to compute the offset from the points only after the transform has been applied. Correct?

Looking at FreeType's `ttgload.c`, that seems to be correct.

**Comment by justvanrossum** [comment]
Although I'm not sure this line is correct:
https://github.com/fonttools/fonttools/blob/c6006a7f8c31cd8e7dddad58894e1ba6c87e1362/Lib/fontTools/ttLib/tables/_g_l_y_f.py#L850

I'm not sure I understand the corresponding FreeType code correctly, but perhaps firstPoint is an index into the point list of the _last_ added glyph, and not the point list as a whole. See the `start_point` argument to `TT_Process_Composite_Component()`.

**Comment by anthrotype** [comment]
I always understood firstPoint as being an index into the whole point list, i.e. the points of all the glyph components added so far

**Comment by anthrotype** [comment]
> I always understood firstPoint as being an index into the whole point list, i.e. the points of all the glyph components added so far

I checked again using ftview and a test font which I made for this purpose and verified that this is indeed the case. firstPt is an index into the points added so far to the composite glyph, secondPt is the index into the list of points of the other glyph that is being added as a component to the current composite glyph.

**Comment by anthrotype** [comment]
@justvanrossum I know it's been 20 years since you wrote this :) but do you remember what you meant by that:

> # XXX Ignoring self.firstPt & self.lastpt for now: I need to implement 
> # something equivalent in fontTools.objects.glyph (I'd rather not 
> # convert it to an absolute offset, since it is valuable information). 

What is `fontTools.objects.glyph`? How do you suggest we implement pen-drawing support for components with firstPt/secondPt nowadays?

I was thinking that, since pen protocol only accepts components with offsets, we compute that firstPt-secondPt and use that as the offset. But maybe you had something more clever in mind, as the comment suggests.

**Comment by justvanrossum** [comment]
With `fontToold.object.glyph` I think I meant something like a proper glyph object, like defcon.Glyph. It sounds like I was pretty much too lazy to figure how to solve this one. I don't think there's a clean way for the pen protocol to handle this. In the meantime: I have never encountered fonts that even used this feature, so I never had any reason to look into it further.

Thinking about it now: I think we should convert it to an offset anyway, and live with the fact that the pen protocol is not capable of round-tripping this rather arcane TT feature.

**Comment by justvanrossum** [comment]
(I think I had (beginnings of) `fontTools.objects.glyph` for a brief while, but for decided it was out of scope for fonttools. It didn't turn into something mature outside of fonttools either, though, I think it just died.)
